### ISSUE TITLE
Require `SubTask.order` to be unique within a task to provide clear, reproducible execution orders for subtasks

### ISSUE THREAD

**Issue by jonrkarr** [SED-ML/sed-ml#95]
For reproducible executions, there should be a well-defined order for the execution of subtasks. This is substantially addressed by the `order` attribute. However, the execution order is not well-defined when multiple subtasks have the same `order`.

There's at least a couple ways to address this
- Require (or at least recommend) that `order` be unique within a task
- Define a canonical ordering algorithm which orders subtasks based on one of
  - The order of their definition in the SED-ML file or 
  - Their task attribute (e.g., alphabetical sorting of `subtask.task.id`)

I propose requiring `order` to unique within a task. This is straightforward to understand and easy to validate. In my opinion, using a canonical ordering algorithm is more complicated, both to understand and implement.

**Comment by matthiaskoenig** [comment]
I agree with making the order unique. 

I don't think a canonical order is necessary (or I don't understand what this should be good for). Multiple subtasks can be executed concurrently using multiprocessing (if there are no dependencies) so such a canonical order would not imply execution order.

**Comment by jonrkarr** [comment]
If one desires to execute sub-tasks in parallel, I think more clarity is needed. To me, the order attribute suggest that sub-tasks are intended to be executed in serial. Two subtasks with the same value of the order attribute, for example, could be used to indicate that those sub-tasks are intended to be independent and could be executed in parallel.

A simpler way to facilitate parallel computation could be for each sub-task to indicate the other sub-tasks that it must follow (its dependencies). This would provide modelers the ability to control dependencies.

**Comment by luciansmith** [comment]
Given that implementation of this is likely to use the 'sort' function newly added to libsedml, I don't see any need to require 'order' attribute values to be unique.  Nobody's going to get anything wrong if the list is sorted in order value.

As I understand it, this comes down to the following suggested addition to the L1v4 spec:

"If the order is not defined or is identical to another subtask, 'executed in any order' includes the ability to execute them in parallel, if desired."

**Comment by matthiaskoenig** [comment]
I agree. Same order states parallelization is possible between subtasks.

**Comment by jonrkarr** [comment]
I don't we think we should rely on everyone using a specific function in libSED-ML. Already, some people are using jLibSED-ML instead.

Text similar to what Lucian wrote would help clarify the intended interpretation of the order attribute.

If subtasks are executed starting from the endpoint of a previous subtask, what is intended to happen when to subtasks with higher values of their order attribute? It seems to me that simulation programs would need to raise an error. Keeping track of a dependency tree rather than a linear order would allow more flexibility here.

**Comment by luciansmith** [comment]
Having a dependency tree is a reasonable suggestion for L2, but for now, everything would have to be harmonized and all data shared between 'order' values.

**Comment by luciansmith** [comment]
To follow up from the meeting yesterday, I finally thought of a way to describe the dependency tree with respect to the 'order' attribute:

* Every subtask is dependent on every other subtask with a lower 'order' value than its own.
* Every subtask is independent of every other subtask with the same or higher 'order' value as its own.
* Every subtask with no 'order' attribute is completely independent of every other subtask.

There will be many dependency trees that will be impossible to construct with this scheme, but it is at least true that an explicit dependency tree can be constructed from these rules.

**Comment by jonrkarr** [comment]
When multiple sub-tasks have the same order, the semantics of how their result should be applied to the input of sub-tasks with higher order is under-defined (e.g., should their result be summed? averaged? differenced?). As a result, one of two additional rules is needed:
- It should not be possible to define sub-tasks with higher order than the lowest order that has multiple sub-tasks.
- All tasks with higher order than the lowest order that has multiple sub-tasks should be executed independently. 

The latter rule is just straw argument. While it outlines the semantics, this would be non-intuitive. Therefore, this should not be the rule.

**Comment by luciansmith** [comment]
There is no concept in SED-ML of 'the result should be applied to the input'.  This does not happen, ever.  Nothing is ever summed; nothing is ever averaged; nothing is ever differenced while executing subtasks.  There is only model state.  You perform a subtask, and any given model state might change, or it might not.  The SED-ML interpreter doesn't need to worry about it at all.

We did just vote to change this slightly with the addition of ComputeChange elements to SubTasks, but even there, it does not have access to anything other than the current model state.  In the past, it was required to take the model state unchanged, however it was received from other subtasks.  Now it has the ability to modify the model state if it wishes.  But again, it cannot sum anything nor average anything nor difference anything that is not part of the static model state that it is given (i.e. variables and current values).

**Comment by matthiaskoenig** [comment]
Many of the constraints we discussed yesterday and also here in the issue become relevant when we move to a more "workflow" like SED-ML system in level 2. At the moment things which can happen with tasks are very limited. All the concepts of chaining simulations or using outputs of one task as input of another task do not really exist at the moment.

**Comment by jonrkarr** [comment]
When there are multiple sub-tasks with the same order, what is the starting model state for sub-tasks with higher order? This is the "output" that I'm referring to. Which model state from previous sub-tasks should be used?

**Comment by luciansmith** [comment]
Oh!  I think I finally understand the question!  Phew.

Let's say that a given RepeatedTask involves two models, mod1 and mod2.  Subtask A performs a timecourse on mod1, and subtask B performs a timecourse on mod2.  So far, so good:  it clearly doesn't matter which order A and B are performed, an indeed they could be executed in parallel if need be, so they are both given an 'order' of '1'.

However, now we have subtask C, which calculates the Jacobian of mod1.  As such, it must have a higher order, so we give it an 'order' of 2.

But!  For the first time, it suddenly matters if we performed things in parallel!  Because the RepeatedTask doesn't necessarily know anything about what A and B are doing; it just hands them each a plate with the model states of mod1 and mod2 on it, and lets them change things.  If you run A and then B, or if you run B and then A, you're fine:  subtask C will still be handed mod1 in the state it was in after A ran.  But!  If you run A and B in parallel, each change the 'suite' of models on the plate differently: mod1 and not mod2 is changed after A runs, and mod2 and not mod1 is changed after B runs.  When it's time to execute C, how to you synchronize your suite again?

This does indeed need clarification in the spec.  For me, my first instinct is to say that when you run things in parallel, you must collect a set of 'diffs' (essentially) on the model states, and apply them all to the set of initial model states.  In this example, we'd apply the A diff to the suite, and mod1 would change, then apply the B diff to the suite, and mod2 would change.   Then you'd run C, and everything will work.

But what do you do if A and B both make changes to the same element of mod1?  The options would seem to be:
1) Error out.
2) Pick one diff arbitrarily and warn the user.
3) Pick one diff arbitrarily and silently continue.
4) Try to sum/average the diffs

Number 1 is obviously the safest, but I can imagine some scenarios where it would not be appropriate:
* If there were no more subtasks to perform anyway, and the repeated task was set 'reset=true'.
* If any subsequent subtasks (or repeated tasks) behaved the same regardless of which diff was applied.

which might indicate that 2 or 3 would be OK, at the tool's discretion.

I think 4) is a fool's errand, and should never be attempted.


My upshot:  tell simulators that when executing subtasks in parallel, they must combine any changes those subtasks make on the model states, unless both a) there are no more subtasks, and b) the repeated task is set 'reset=true'.

**Comment by jonrkarr** [comment]
For L1V4, I suggest scenario (1).

For L2, I suggest we change the way that dependencies among sub-tasks are specified to allow more flexibility. The order attribute largely only supports a linear flow of sub-tasks (with parallelization at individual orders). If SED-ML instead captured the dependency of each sub-task (i.e., id of the sub-task that it has to run after), SED-ML could support more complex trees of sub-tasks. Such trees would cover scenarios (2) and (3) by allowing the modeler to explicitly state the dependencies they want. If someone wanted to do (4), this could be encoded into a chain of sub-tasks, albeit opaquely; arguably this could be done already, although not with parallelization.

Such dependencies trees would make SED-ML increasingly look like a workflow language. For example, such sub-task dependencies could be specified similar to how SLURM, PBS, and GitHub Actions capture dependencies between jobs.

As part of the L2, we could try to capture merging of "results" of sub-tasks directly in SED-ML rather than something that has to be encoded into a simulation task. To think about how to capture this, it would be helpful to see a few concrete examples. Its seems that specific simulations may want to have sub-tasks merged in different ways. Its not clear to me (yet) what should be captured in a more general way.

**Comment by matthiaskoenig** [comment]
> If you run A and B in parallel, each change the 'suite' of models on the plate differently: mod1 and not mod2 is changed after A runs, and mod2 and not mod1 is changed after B runs.

I don't see this as an issue
It is clear from the definition that if you run a task with higher order all lower order tasks must already have been executed and the models are in the state of the end execution of all lower tasks.
I.e. as soon as C (oder 2) runs, mod1 has the end stage of A executed and mod2 has the end stage of B executed. It does not matter at all if you run A and then B, B and then A or A and B at the same time, because the two tasks change different models!

You cannot parallelize two tasks with the same order in the listOfSubTasks if they change the same model and if the reset is False on later subtasks using the model. There is no way to know what the end stage of the model is. This should be clearly written in the specification.

**Comment by jonrkarr** [comment]
I agree that if subtasks with the same order change the states of different models, the semantics of their parallel execution of these subtasks is clear. Where is this constraint that sub-tasks that have the same order must reference different models described in the specifications?

I don't see how `resetModel` is relevant here. My interpretation is that `resetModel` applies to the whole iteration of a repeated task, not between individual subtasks.

**Comment by luciansmith** [comment]
I have added the following to address this issue.  First, in the listOfSubTasks section:

"In some cases, it may be possible to run some subtasks in parallel if the following conditions are met:
 The order attribute of both subtasks must be the same or unset.
 The subtasks must not change the model state of the same model.
 If there are subtasks to run with a higher order value, or if the resetModel attribute of the
RepeatedTask is \false", it must be possible to obtain and apply the changed model states from
all parallel subtasks before running the next subtask."

and then in the 'order' section:

"Leaving the order undefined for a SubTask implies that the SubTask may be executed before, after, or
in parallel with any other SubTask. Giving the same order to multiple SubTask elements is an explicit
statement that each SubTask in the group may be executed before, after, or in parallel with any other
SubTask in the group. It is recommended that users always explicitly set the order attribute for this
reason."

Leaving this open to ensure that the wording is OK and that nothing else is missing (or wrong!)

**Comment by jonrkarr** [comment]
I think these conditions are more restrictive than necessary. I would separate two issues:

(a) whether a description of a set of subtasks is valid (i.e., semantically well-defined) 
(b) whether subtasks can be executed in parallel

Regarding (a), subtasks are ill-defined if (i) a there a multiple subtasks with the same order that make (different) changes to the same model and (ii) there are additional subtask(s) with higher order that involve the same model. In this case, its semantically unclear what model to use for the later subtask(s). In my opinion, a validation tool should identify such cases and raise errors.

Assuming simulation tools raises errors when (a) is true, subtasks with the same order could always be run in parallel. In addition, subtasks with successive orders that don't share models or that don't share models can also be executed in parallel.

**Comment by luciansmith** [comment]
Hmm, I think you're right that the issues need to be separated.

For a), I think that it would actually be valid for two tasks to change the same model and have the same order if they could be executed in either order (i.e. if 'A then B' and 'B then A' both resulted in the same model state).  However, this would be a situation that could not be parallelized.  Do we want to say that it is therefore illegal?

For b), if you want to relax the restrictions, perhaps the best thing to say is simply 'if you analyze the tasks and determine they can be parallelized, then go for it'.  This would allow people to ignore the 'order' if they could analytically prove that the order didn't matter, and would also save them from the situation in a).

(As an extra note:  if task1 modifies model1 and task2 modifies model2, that's insufficient to know if they can be parallelized;  one could be parameterized with the other, i.e. with a Variable for task1 that targets model2.)

**Comment by jonrkarr** [comment]
For (a), I think two tasks could have the same order and make different changes to the same model so long as there are no subtasks with higher orders. In this scenario, the subtasks could be executed in any order or in parallel.

I think the conditions I outlined describe when subtasks can be executed in parallel.
- Subtasks with the same order can always be executed in parallel (because of the validation implied above)
- Sequences (by increasing order) of subtasks which do not make different changes to the same model (e.g., subtasks which make no changes, which change different models, or which make the same changes to the same models).

These conditions would support parallelism in the example you describe

> task1 modifies model1 and task2 modifies model2

**Comment by luciansmith** [comment]
As in https://github.com/SED-ML/sed-ml/issues/54:

The latest updates to the spec call out the possibility of parallelization in a few places.  In listOfTasks:

"Each top-level task is defined such that its execution is independent of the others: if one task is executed
after another, the states of the models must be completely reset so there’s no cross-contamination of one
task to the next. This means that the top-level tasks are particularly well suited to being executed in
parallel, should that be desired."

in 'resetModel':

"When the resetModel attribute is set to "true", the individual repeats of the task may be paralellizable,
assuming any child Range of the RepeatedTask does not depend on the results of any individual repeat
(as is theoretically possible for the FunctionalRange, but not for the other Range types.)"

in 'listOfSubmodels':

"In some cases, it may be possible to run some subtasks in parallel. Interpreters may use the order
attribute as a hint in making the decision about what to parallelize, but in general, should be prepared
to examine which Model each SubTask is modifying and which model or models are being used as input."

(I changed to this from my previous list of requirements, which was too restrictive.  Instead, I'm basically just punting to interpreters to determine this on their own.)

in 'order':

"Leaving the order undefined for a SubTask implies that the SubTask may be executed before or after
any other SubTask. Giving the same order to multiple SubTask elements is an explicit statement that
each SubTask in the group may be executed before or after any other SubTask in the group. It is
recommended that users always explicitly set the order attribute for this reason.

Any order value does not imply whether the SubTask may be executed in parallel with other SubTask
elements. Interpreters who wish to parallelize subtasks should operate from the assumption that in the
default case, each SubTask would be executed in some order, and adjust accordingly."

**Comment by jonrkarr** [comment]
> When the resetModel attribute is set to "true", the individual repeats of the task may be paralellizable,
assuming any child Range of the RepeatedTask does not depend on the results of any individual repeat
(as is theoretically possible for the FunctionalRange, but not for the other Range types.)

This raises a question about when functional range is intended to be evaluated. My interpretation is that the range if evaluated before an iterations of run. I'm not 100% sure, but I think jLibSED-ML works this way too.

> In some cases, it may be possible to run some subtasks in parallel. Interpreters may use the order
attribute as a hint in making the decision about what to parallelize, but in general, should be prepared
to examine which Model each SubTask is modifying and which model or models are being used as input.

I think this can be simplified:

"Subtasks can also be executed in parallel when they do not share any state. Interpreters can determine this from the descriptions of the subtasks."

**Comment by luciansmith** [comment]
Oh, hmm.  I guess I don't know for sure either way when FunctionalRanges are supposed to be evaluated.  I was remembering when they were sort of proposed to have 'while' functionality (instead of 'for'), but then that didn't quite work.. or at least we thought it didn't work?

Let's say that they're evaluated at the beginning of the run, and that's probably simplest, not to mention the most likely to have implementations.

Also, I like your changes; I made the following changes:

Added to FunctionalRange:

"The value of any Variable child of a FunctionalRange should be calculated before the first simulation
begins, and will not be affected by any SubTask in the RepeatedTask."

'resetModel':

"When the resetModel attribute is set to "true", the individual repeats may be executed in parallel."

And then I used your text for 'listofSubTasks'
